modules/optimization/src/utils/ParameterMesh.C

```diff
@@ -429,67 +441,127 @@ ParameterMesh::computeGradientL2RegularizationObjective(
     std::unique_ptr<FEBase> fe(FEBase::build(dim, _fe_type));
     fe->attach_quadrature_rule(&qrule);
 
-    // Request derivatives before reinit
+    // Request shape functions and derivatives before reinit
     const std::vector<Real> & JxW = fe->get_JxW();
+    const std::vector<std::vector<Real>> & phi = fe->get_phi();
     const std::vector<std::vector<RealGradient>> & dphi = fe->get_dphi();
 
     // Reinitialize for current element
     fe->reinit(elem);
 
     // Loop over quadrature points
     for (const auto qp : make_range(qrule.n_points()))
+      compute_func(phi, dphi, qp, dof_indices, JxW, result);
```

Comment:
Why not just call the lambda here instead of passing it.  Maybe each regularization should be a private struct in this class with their own compute_func.  I understand what you are doing and this is what we asked for, I just didn't realize it would look like this.  I'm fine with this now, I understand it enough to implement new regularizations.  We can refactor it some day if we have a better design.